### PR TITLE
feat: make detection of crd spec/status more flexible

### PR DESCRIPTION
**Pull request by iocanel**
-------

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [ ] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

File: crd-generator/apt/src/main/java/io/fabric8/crd/generator/apt/CustomResourceAnnotationProcessor.java [hunk 1]
```diff
@@ -160,4 +162,34 @@ public void close() throws IOException {
...
+
+  private static Optional<ClassRef> getCustomResourceStatus(TypeDef typeDef) {
+    return getCustomResourceRef(typeDef)
+      .filter(c -> c.getArguments() != null && c.getArguments().size() == 2)
+      .map(c -> c.getArguments().get(1))
+      .filter(t -> t instanceof ClassRef)
+      .map(t -> (ClassRef) t);
+  }
+
+  public static Stream<ClassRef> superClassStream(TypeDef typeDef) {
```
Comment (metacosm) [inline crd-generator/apt/src/main/java/io/fabric8/crd/generator/apt/CustomResourceAnnotationProcessor.java:188]: I was thinking about something like this but wasn't too sure how to go about it…

File: crd-generator/apt/src/main/java/io/fabric8/crd/generator/apt/CustomResourceAnnotationProcessor.java [hunk 2]
```diff
@@ -160,4 +162,34 @@ public void close() throws IOException {
...
     }
   }
+
+  private static Optional<ClassRef> getCustomResourceRef(TypeDef typeDef) {
+    return superClassStream(typeDef)
+      .filter(s -> s.getDefinition().getFullyQualifiedName().equals(CustomResource.class.getName()))
+      .findFirst();
+  }
+
+  private static Optional<ClassRef> getCustomResourceSpec(TypeDef typeDef) {
```
Comment (metacosm) [inline crd-generator/apt/src/main/java/io/fabric8/crd/generator/apt/CustomResourceAnnotationProcessor.java:172]: Maybe only one method with the index of the argument as parameter?

File: crd-generator/apt/src/main/java/io/fabric8/crd/generator/apt/CustomResourceAnnotationProcessor.java [hunk 3]
```diff
@@ -29,16 +29,26 @@
...
 import io.fabric8.kubernetes.model.annotation.Version;
 import io.sundr.codegen.CodegenContext;
+import io.sundr.codegen.Constants;
+import io.sundr.codegen.functions.ClassTo;
 import io.sundr.codegen.functions.ElementTo;
+import io.sundr.codegen.model.ClassRef;
 import io.sundr.codegen.model.TypeDef;
 import io.sundr.codegen.model.TypeRef;
+import io.sundr.codegen.utils.TypeUtils;
+import sun.rmi.rmic.iiop.SpecialClassType;
```
Comment (metacosm) [inline crd-generator/apt/src/main/java/io/fabric8/crd/generator/apt/CustomResourceAnnotationProcessor.java:39]: Wrong import

**Review by metacosm — CHANGES_REQUESTED** [review]

**Comment by metacosm** [comment]
Sorry didn't think that merging #2909 would automatically close this one as well.

**Comment by iocanel** [comment]
> Sorry didn't think that merging #2909 would automatically close this one as well.

Me neither! No worries. I created a fresh one!

**Pull request closed without merging**